shpc/client/upgrade.py

```diff
@@ -0,0 +1,161 @@
+        if args.dryrun:
+            version_info = upgrade(args.upgrade_recipe, cli, args, dryrun=True) # This returns the latest version if its available, else returns None 
+            if version_info:
+                logger.info(f"You do not have the latest version installed.\nLatest version avaiable is {version_info}")
+            else:
+                logger.info(f"You have the latest version of {args.upgrade_recipe} installed.")
+
+        # Upgade the software
+        else:
+            upgrade(args.upgrade_recipe, cli, args, dryrun=False, force=args.force)
+
+    # Upgrade all installed software
+    elif args.upgrade_all:
+        # Store the number of all outdated software
+        num_outdated = 0
+
+        # Does the user just want a dry-run of all software?
+        if args.dryrun:
+            print("Performing a dry-run on all your software...")
+            for software in installed_software.keys():
+                version_info = upgrade(software, cli, args, dryrun=True)
+                if version_info:
+                    logger.info(f"{software} is outdated. Latest version available is {version_info}")
+                    num_outdated += 1
+                else:
+                    logger.info(f"{software} is up to date.")
+            # Provide a report on the dry-run
+            if num_outdated == 0:
+                logger.info("All your software are currently up to date.")
+            else:
+                logger.info(f"You have a total of {num_outdated} outdated software.")
+
+        # Upgrade all software
+        else:
+            print("Checking your list to upgrade outdated software...")
+            for software in installed_software.keys():
+                # Attempt upgrade on each software
+                upgrade_info = upgrade(software, cli, args, dryrun=False, force=args.force)
+                # Count actual upgrades
+                if upgrade_info:
+                    num_outdated += 1
+            if num_outdated == 0:
+                logger.info("No upgrade needed. All your software are up to date.")
+            else:
+                logger.info(f"Updated {num_outdated} outdated software from your list\nAll your software are now up to date.")                    
+
+
+def upgrade(name, cli, args, dryrun=False, force=False):
+    """
+    Upgrade a software to its latest version. Or preview available upgrades from the user's software list
+    """
+    # Add namespace 
+    name = cli.add_namespace(name)
+    
+    # Load the container configuration for the specified recipe
+    config = cli._load_container(name)
+
+    #Store the installed versions and the latest version tag
+    installed_versions = cli.list(pattern=name, return_modules=True)
+    latest_version_tag = get_latest_version(name, config)
+
+    # Compare the latest version with the user's installed version
+    if any(latest_version_tag in versions for versions in installed_versions.values()):
```

Comment:
Is this equivalent ?
```suggestion
    if latest_version_tag in installed_versions.values():
```

Reply:
No, it's not, because installed_versions is a dictionary whose values are a set of versions. So the logic I used was to extract each version from the set

Reply:
Got it, thanks !